### DISCUSSION TITLE
String on stack

### DISCUSSION THREAD

**Discussion by kefyru** [dotnet/runtime#78509] [Q&A] [answered]
I have made it. Now i have question. How dangerous is it.
I understand, that such strings cant be assigned to object fields, to array elements e.t.c., we should use it like span.
So question is:
Are all platforms uses string header 12 bytes (4 methodTable + 4 ??? + 4 length)?
Are all platforms uses null-terminator `'\0'` at the end?
What about AOT compilation?
Am i miss something?

```cs
var dictionary = new Dictionary<string,string>();
dictionary["hello"] = "world";
var methodTable = *(int*)*(void**)Unsafe.AsPointer(ref Unsafe.AsRef(string.Empty));
ReadOnlySpan<char> key = "hello";
var buffer = stackalloc int[(key.Length >> 1) + 3 + 1]; 
buffer[0] = methodTable;
buffer[1] = 1; // i cant find info what is it, thinking 1 = pinned;
buffer[2] = key.Length;
var span = new Span<char>(buffer + 3, key.Length + 1);
key.CopyTo(span);
span[key.Length] = '\0';
ref var strKeyOnStack = ref Unsafe.AsRef<string>(&buffer);

if (dictionary.TryGetValue(strKeyOnStack, out var v) && v == "world") Console.WriteLine("Now we can create strings on stack");
```

**Comment by kefyru** [comment]
Works in blazor webassembly

**Comment by OwnageIsMagic** [comment]
Methodtable is pointer (not fixed 4 bytes), and layout it's not guaranteed and very likely to change (i.e. it was changed on net35->net40, CoreCLR have some internal switches to sometimes omit object header), so without exact runtime version assert it's not safe.
And if you leak this object (accidentally add it to dictionary as key or value) you will corrupt process.

**Reply by OwnageIsMagic** [reply]
and you should probably use `byte` buffer, not `int`

**Comment by teo-tsirpanis** [comment]
> How dangerous is it.

_Extremely_ dangerous, can easily corrupt the runtime in unpredictable ways (and if it works on your machine there's absolutely no guarantee that it will continue to work in another machine or in a future .NET version), just forget about it, or wait for #2010.

**Comment by jkotas** [comment, marked as answer]
If the GC runs while reference to one of these strings is on the stack, it is very likely going to crash.

**Reply by OwnageIsMagic** [reply]
I'm sorry, could you elaborate more why it problem for GC? As I understand it would be treated just as an ordinary GC root and won't be marked for collection and it doesn't contain any managed fields.

**Reply by teo-tsirpanis** [reply]
The GC assumes that objects are in its heaps. Not the stack.

**Reply by OwnageIsMagic** [reply]
But why does it cares as long as the object is not eligible for collection?

**Reply by teo-tsirpanis** [reply]
Objects must be in heaps managed by the GC because it might need to move them when doing a compacting collection, and generally to keep track of them. It is not currently designed to handle objects in the stack, and when .NET supports escape analysis, it will be a transparent optimization.

**Reply by OwnageIsMagic** [reply]
Ahh, generation promoting, that what I missed. But why GC bothers with ref vars? There is no way that ref var would be the last reference to object, so GC can skip it as it's guaranteed that it can find this object trough another root.

**Reply by teo-tsirpanis** [reply]
> There is no way that ref var would be the last reference to object

Why not? If you write `Span<byte> x = new byte[10];` the array's only reference would be through the span.

**Reply by jkotas** [reply]
The ref var in your example is not the problem. The problem is when that ref var is passed to `TryGetValue`. It will end up with ordinary `string` local pointing to stack.

**Reply by OwnageIsMagic** [reply]
@teo-tsirpanis it's late night there, seems like I need some sleep..
@jkotas It's not my example, just trying clear some points for myself.
Thank you very much for your time.
And one last question. How GC manages to keep alive this object, while there is no direct reference to it? Does it just searching for nearest object in heap pointed by `ref i`?
```c#
ref var i = ref Get();
++i;
GC.Collect(3, GCCollectionMode.Forced, blocking: true, compacting: true);
++i;
Console.WriteLine(i);
   
static ref int Get() => ref new C().fi;
class C { public int fi; }
```

**Reply by jkotas** [reply]
`ref var i` is interior pointer. Interior pointers are allowed into stack, unmanaged memory or GC heap. When they point into GC heap, they keep the containing object alive.

**Reply by OwnageIsMagic** [reply]
Thanks for pointing out right words. Found my answer here [Recover containing GC object from managed 'ref' interior pointer](https://stackoverflow.com/questions/52824792/recover-containing-gc-object-from-managed-ref-interior-pointer)
Thanks again for a great job you guys are doing here.
